### ISSUE TITLE
duplicate symbols in OS X build

### ISSUE THREAD

**Issue by jowens** [gunrock/gunrock#20]
Just wondering if this has come up in any other builds.

````
[ 18%] Building CXX object gunrock/CMakeFiles/gunrock.dir/__/externals/moderngpu/src/mgpuutil.cpp.o
Linking CXX shared library ../lib/libgunrock.dylib
duplicate symbol __ZN7gunrock4util12stringprintfEPKcz in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/bfs/gunrock_generated_bfs_app.cu.o
duplicate symbol __ZN7gunrock7graphio5SprngEv in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/bfs/gunrock_generated_bfs_app.cu.o
duplicate symbol __ZN7gunrock7graphio4FlipEv in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/bfs/gunrock_generated_bfs_app.cu.o
duplicate symbol __ZN7gunrock7graphio10VaryParamsEPdS1_S1_S1_ in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/bfs/gunrock_generated_bfs_app.cu.o
duplicate symbol __ZN7gunrock4util12stringprintfEPKcz in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/bc/gunrock_generated_bc_app.cu.o
duplicate symbol __ZN7gunrock7graphio5SprngEv in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/bc/gunrock_generated_bc_app.cu.o
duplicate symbol __ZN7gunrock7graphio4FlipEv in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/bc/gunrock_generated_bc_app.cu.o
duplicate symbol __ZN7gunrock7graphio10VaryParamsEPdS1_S1_S1_ in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/bc/gunrock_generated_bc_app.cu.o
duplicate symbol __ZN7gunrock4util12stringprintfEPKcz in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/cc/gunrock_generated_cc_app.cu.o
duplicate symbol __ZN7gunrock7graphio5SprngEv in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/cc/gunrock_generated_cc_app.cu.o
duplicate symbol __ZN7gunrock7graphio4FlipEv in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/cc/gunrock_generated_cc_app.cu.o
duplicate symbol __ZN7gunrock7graphio10VaryParamsEPdS1_S1_S1_ in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/cc/gunrock_generated_cc_app.cu.o
duplicate symbol __ZN7gunrock4util12stringprintfEPKcz in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/pr/gunrock_generated_pr_app.cu.o
duplicate symbol __ZN7gunrock7graphio5SprngEv in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/pr/gunrock_generated_pr_app.cu.o
duplicate symbol __ZN7gunrock7graphio4FlipEv in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/pr/gunrock_generated_pr_app.cu.o
duplicate symbol __ZN7gunrock7graphio10VaryParamsEPdS1_S1_S1_ in:
    CMakeFiles/gunrock.dir/app/topk/gunrock_generated_topk_app.cu.o
    CMakeFiles/gunrock.dir/app/pr/gunrock_generated_pr_app.cu.o
ld: 16 duplicate symbols for architecture x86_64
````

**Comment by jowens** [comment]
@yzhwang I actually don't know how this code compiles anywhere. If I look at gunrock/graphio/utils.cuh, for instance, I see that there's not only a declaration but a /definition/ of a function (in this case, Sprng). So anyplace this header is included will result in a Sprng in that compilation unit (foo.o), and then linking foo.o and bar.o will have two Sprng functions, which is bad. My understanding is common practice mandates no definitions in a header file; definitions need to go in a cpp/cxx/cu file and be linked. #pragma once does not solve the linking problem across header files.

**Comment by yzhwang** [comment]
I will assgin this to @ydwu since he is the guy who added the code. Yuduo, can you look into the Sprng function problem here? Thanks!

**Comment by yzhwang** [comment]
I think the reason we haven't noticed this problem is that on linux we haven't used cmake to build. But this needs to be fixed for users who use cmake.

**Comment by jowens** [comment]
It's all util and graphio functions.

**Comment by jowens** [comment]
It's pretty ugly, but I think inlining the functions in question would temporarily solve the problem (not a great solution, but would work). That's util/{stringprint,Sprng} and graphio/{Flip,VaryParams}, as best I can tell.

**Comment by jowens** [comment]
Fixed in 25b4a2636431663806243a255d35188049293216 - thanks!
